apollo-federation/src/operation/contains.rs

```diff
@@ -41,6 +42,95 @@ fn same_value(left: &executable::Value, right: &executable::Value) -> bool {
+/// hashing directives in a way consistent with [same_directives()].
+///
+/// Note that Floats and Ints are compared textually and not parsed numerically. This is fine for
+/// the purposes of hashing. For object comparison semantics, see [compare_sorted_object_pairs()].
+fn compare_sorted_value(left: &executable::Value, right: &executable::Value) -> std::cmp::Ordering {
+    use apollo_compiler::executable::Value;
+    fn discriminant(value: &Value) -> u8 {
+        match value {
+            Value::Null => 0,
+            Value::Enum(_) => 1,
+            Value::Variable(_) => 2,
+            Value::String(_) => 3,
+            Value::Float(_) => 4,
+            Value::Int(_) => 5,
+            Value::Boolean(_) => 6,
+            Value::List(_) => 7,
+            Value::Object(_) => 8,
+        }
+    }
+    match (left, right) {
+        (Value::Null, Value::Null) => std::cmp::Ordering::Equal,
+        (Value::Enum(left), Value::Enum(right)) => left.cmp(right),
+        (Value::Variable(left), Value::Variable(right)) => left.cmp(right),
+        (Value::String(left), Value::String(right)) => left.cmp(right),
+        (Value::Float(left), Value::Float(right)) => left.as_str().cmp(right.as_str()),
+        (Value::Int(left), Value::Int(right)) => left.as_str().cmp(right.as_str()),
+        (Value::Boolean(left), Value::Boolean(right)) => left.cmp(right),
+        (Value::List(left), Value::List(right)) => left.len().cmp(&right.len()).then_with(|| {
+            left.iter()
+                .zip(right)
+                .map(|(left, right)| compare_sorted_value(left, right))
+                .find(|o| o.is_ne())
+                .unwrap_or(std::cmp::Ordering::Equal)
+        }),
+        (Value::Object(left), Value::Object(right)) => compare_sorted_name_value_pairs(
+            left.iter().map(|pair| (&pair.0, &pair.1)),
+            right.iter().map(|pair| (&pair.0, &pair.1)),
+        ),
+        _ => discriminant(left).cmp(&discriminant(right)),
+    }
+}
+
+/// Compare the (name, value) pair iterators, which are assumed to be sorted by name and have sorted
+/// values. This is used for hashing objects/arguments in a way consistent with [same_directives()].
+///
+/// Note that pair iterators are compared by length, then lexicographically by name, then finally
+/// recursively by value. This is intended to compute an ordering quickly for hashing.
+fn compare_sorted_name_value_pairs<'doc>(
+    left: impl ExactSizeIterator<Item = (&'doc Name, &'doc Node<executable::Value>)>,
+    right: impl ExactSizeIterator<Item = (&'doc Name, &'doc Node<executable::Value>)>,
+) -> std::cmp::Ordering {
+    left.len().cmp(&right.len()).then_with(|| {
+        let mut pairs = vec![];
+        for ((left_name, left_value), (right_name, right_value)) in left.zip(right) {
+            match left_name.cmp(right_name) {
+                std::cmp::Ordering::Equal => pairs.push((left_value, right_value)),
```

Comment:
Hmm, we are preparing this `pairs` vec with `(left_value, right_value)`, but then afterwards we just iterate through it in order and compare each `left_value` with `right_value`. Can't we just do the `compare_sorted_value` inline here?

Reply:
@goto-bus-stop 
My thoughts here were that recursing into the values with `compare_sorted_value()` could potentially be expensive, but differences in name/key sets between the pairs would probably be faster to compute, so we should early return if the key sets are different and only recurse if they're the same.

It looks like I can avoid the `vec` creation by just having separate iterators for names and values, so I went ahead and changed the API of `compare_sorted_name_value_pairs()` accordingly in the latest commit, let me know if there's a simpler way and I can make a followup PR.